### PR TITLE
Catch panic.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by chriseth** [comment]
If we want to support arbitrary error codes, we have to implement `tryabidecode`, i.e. an abi decoder that returns a status flag instead of reverting.

File: test/libsolidity/syntaxTests/tryCatch/double_panic.sol [hunk 1]
```diff
@@ -0,0 +1,11 @@
+contract C {
+    function f() public {
+        try this.f() {
+        } catch Panic(bytes memory) {
+        } catch Panic(uint) {
+        }
+    }
+}
```
Comment (mijovic) [inline test/libsolidity/syntaxTests/tryCatch/double_panic.sol:8]:
```suggestion
}
// ====
// EVMVersion: >=byzantium
```

File: test/libsolidity/syntaxTests/tryCatch/double_panic_correct.sol [hunk 2]
```diff
@@ -0,0 +1,10 @@
+contract C {
+    function f() public {
+        try this.f() {
+        } catch Panic(uint) {
+        } catch Panic(uint) {
+        }
+    }
+}
+// ----
```
Comment (mijovic) [inline test/libsolidity/syntaxTests/tryCatch/double_panic_correct.sol:11]:
```suggestion
// ====
// EVMVersion: >=byzantium
// ----
```

File: test/libsolidity/syntaxTests/tryCatch/error_and_panic.sol [hunk 3]
```diff
@@ -0,0 +1,9 @@
+contract C {
+    function f() public {
+        try this.f() {
+        } catch Error(string memory) {
+        } catch Panic(uint) {
+        }
+    }
+}
+// ----
```
Comment (mijovic) [inline test/libsolidity/syntaxTests/tryCatch/error_and_panic.sol:11]:
```suggestion
// ====
// EVMVersion: >=byzantium
// ----
```

File: test/libsolidity/syntaxTests/tryCatch/wrong_panic.sol [hunk 4]
```diff
@@ -0,0 +1,9 @@
+contract C {
+    function f() public {
+        try this.f() {
+        } catch Panic() {
+        }
+    }
+}
+// ----
```
Comment (mijovic) [inline test/libsolidity/syntaxTests/tryCatch/wrong_panic.sol:10]:
```suggestion
// ====
// EVMVersion: >=byzantium
// ----
```

File: test/libsolidity/syntaxTests/tryCatch/wrong_panic_2.sol [hunk 5]
```diff
@@ -0,0 +1,9 @@
+contract C {
+    function f() public {
+        try this.f() {
+        } catch Panic(bytes memory) {
+        }
+    }
+}
+// ----
```
Comment (mijovic) [inline test/libsolidity/syntaxTests/tryCatch/wrong_panic_2.sol:10]:
```suggestion
// ====
// EVMVersion: >=byzantium
// ----
```

File: libsolidity/analysis/TypeChecker.cpp [hunk 6]
```diff
@@ -1040,26 +1041,46 @@ void TypeChecker::endVisit(TryStatement const& _tryStatement)
...
+				)
+					m_errorReporter.typeError(1271_error, clause.location(), "Expected `catch Panic(uint ...) { ... }`.");
+			}
 		}
 		else
 			m_errorReporter.typeError(
 				3542_error,
 				clause.location(),
-				"Invalid catch clause name. Expected either `catch (...)` or `catch Error(...)`."
+				"Invalid catch clause name. Expected either `catch (...)`, `catch Error(...)` or `catch Panic(...)`."
```
Comment (mijovic) [inline libsolidity/analysis/TypeChecker.cpp]: Maybe using oxford comma?
```suggestion
				"Invalid catch clause name. Expected either `catch (...)`, `catch Error(...)`, or `catch Panic(...)`."
```

File: libsolidity/codegen/ContractCompiler.cpp [hunk 7]
```diff
@@ -971,31 +971,46 @@ bool ContractCompiler::visit(TryStatement const& _tryStatement)
...
 			solAssert(false, "");
 
-	solAssert(_catchClauses.size() == 1ul + (structured ? 1 : 0) + (fallback ? 1 : 0), "");
+	solAssert(_catchClauses.size() == 1ul + (error ? 1 : 0) + (panic ? 1 : 0) + (fallback ? 1 : 0), "");
 
 	evmasm::AssemblyItem endTag = m_context.newTag();
 	evmasm::AssemblyItem fallbackTag = m_context.newTag();
-	if (structured)
+	evmasm::AssemblyItem panicTag = m_context.newTag();
+	if (error != nullptr || panic != nullptr)
```
Comment (mijovic) [inline libsolidity/codegen/ContractCompiler.cpp]:
```suggestion
	if (error || panic)
```

Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp]: It was weirdly failing before I changed it, I'll try to revert again.

**Comment by mijovic** [comment]
If tests are passing just needs squash and should be ready

**Comment by chriseth** [comment]
I would like to add some more tests.

**Comment by chriseth** [comment]
Adding more tests was useful, because it uncovered a bug in the `Error(string)` decoder.

File: libsolidity/codegen/YulUtilFunctions.cpp [hunk 8]
```diff
@@ -4014,21 +4031,39 @@ string YulUtilFunctions::tryDecodeErrorMessageFunction()
 				if gt(length, 0xffffffffffffffff) { leave }
 
 				let end := add(add(msg, 0x20), length)
-				if gt(end, add(data, returndatasize())) { leave }
+				if gt(end, add(data, sub(returndatasize(), 4))) { leave }
```
Comment (chriseth) [inline libsolidity/codegen/YulUtilFunctions.cpp:4034]: This is the bugfix. The bug allows malformed error data to be decoded as correct error data, which could lead to nondeterminism when this data is used.

File: libsolidity/codegen/ContractCompiler.cpp (hunk 7 above)
Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp]: Malformed in the sense that the error data is too short for the length stored inside it.

File: libsolidity/codegen/YulUtilFunctions.cpp (hunk 8 above)
Comment (chriseth) [reply libsolidity/codegen/YulUtilFunctions.cpp:4034]: Malformed in the sense that the error data is too short for the length stored inside it.

**Review by mijovic — APPROVED** [review]

**Comment by chriseth** [comment]
Missed updating the documentation.

**Pull request merged**